designs/dotnetcore-debugging.md

```diff
@@ -0,0 +1,196 @@
+
+1. Set `request: attach` since we are attaching to already running .NET application;
+2. Set `pipeProgram` to `docker` (no need for SSH luckily);
+3. Use `exec -i <container name>` as `pipeArgs` to step inside the running container;
+4. Specify `debuggerPath` as **"/tmp/lambci_debug_files/vsdbg"** (this path reflects current mounting point for `--debugger-path` , refer to this [PR](https://github.com/awslabs/aws-sam-cli/pull/565/files) for more details);
+5. Set `quoteArgs` to `False` and let the Docker accept arguments correctly (without the need for `sh`);
+6. Set `processId` to 1 as Docker always assigns PID 1 to entry point executables.
+7. `sourceFileMap` is used for mapping breakpoints. `/var/task` is the place, where customer's code is being mounted to.
+
+_Examine sample launch.json configuration_
+
+```json
+{
+    "version": "0.2.0",
+    "configurations": [
+        {
+            "name": ".NET Core Docker Attach",
+            "type": "coreclr",
+            "request": "attach",
+            "processId": "1",
+            "pipeTransport": {
+                "pipeProgram": "docker",
+                "pipeArgs": [ "exec -i <container name>"],
+                "debuggerPath": "/tmp/lambci_debug_files/vsdbg",
+                "pipeCwd": "${workspaceRoot}",
+                "quoteArgs": false
+            },
+
+            "sourceFileMap": {
+                "/var/task": "${workspaceRoot}"
+            }
+        }
+    ]
+}
+```
+
+##### Container name caveat
+
+With this `launch.json` in place the only caveat left is the container name to step into. As for now, SAM will not assign consistent names for started containers, and neither it provides the way for the user to supply that name. So my proposal is to create _new flag_ for `sam local` command -  `--container-name` which will provide the user a way to specify the name for the container when it runs. This is a required configuration since user needs consistent naming to have usable attaching experience. SAM will use provided name and add it to `container.start()` configuration.
+
+### Open questions
+
+1. `--container-name` - does this option make sense to you, guys, maybe you have some thoughts of how to it better without exposing the ability to assign the name to the lambda container. As for me, I think this option is more than handy, and does not uncover additional implementation details, as the user is now already well aware, that his function is invoked inside Docker container. Maybe we want to have some environment variable too? Will we include that in `.samrc`? 
+
+2. A bit off-topic, but still, has someone else encountered the problem with python 3.7.1 ? As it now does not flush the `stdout` and `stderr` from the running container immediately but rather after SAM ends the invocation, and user does not see _"waiting for the debugger to attach..."_ message. 
+
+   _From SAM code - container.py:_
+
+   ```python
+    for frame_type, data in output_itr:
```

Comment:
Yeah, I believe this : https://github.com/awslabs/aws-sam-cli/pull/729

We do have unbuffered streams, whenever a debug port is set explicitly. but in this case there is no debugger port. because of the .net debugging works here, so we may need to slightly rethink that.